x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java

```diff
@@ -177,19 +181,64 @@ private void onFatalFailure(AsyncSearchTask task, Exception error, boolean shoul
     private void onFinalResponse(AsyncSearchTask searchTask,
                                  AsyncSearchResponse response,
                                  Runnable nextAction) {
-        store.updateResponse(searchTask.getExecutionId().getDocId(), threadContext.getResponseHeaders(),response,
-            ActionListener.wrap(resp -> unregisterTaskAndMoveOn(searchTask, nextAction),
+        store.updateResponse(searchTask.getExecutionId().getDocId(), threadContext.getResponseHeaders(), response,
+            ActionListener.wrap(
+                resp -> unregisterTaskAndMoveOn(searchTask, nextAction),
                 exc -> {
                     Throwable cause = ExceptionsHelper.unwrapCause(exc);
                     if (cause instanceof DocumentMissingException == false &&
                             cause instanceof VersionConflictEngineException == false) {
                         logger.error(() -> new ParameterizedMessage("failed to store async-search [{}]",
                             searchTask.getExecutionId().getEncoded()), exc);
+                        updateStoredResponseWithFailure(
+                            searchTask.getExecutionId().getDocId(),
+                            response,
+                            exc,
+                            ActionListener.wrap(() ->  unregisterTaskAndMoveOn(searchTask, nextAction))
+                        );
+                    } else {
+                        unregisterTaskAndMoveOn(searchTask, nextAction);
                     }
-                    unregisterTaskAndMoveOn(searchTask, nextAction);
                 }));
     }
 
+    private void updateStoredResponseWithFailure(String docId,
+                                                 AsyncSearchResponse originalASR,
+                                                 Exception updateException,
+                                                 ActionListener<UpdateResponse> listener) {
+        SearchResponse originalSR = originalASR.getSearchResponse();
+        InternalSearchResponse failureISR = new InternalSearchResponse(
+            new SearchHits(SearchHits.EMPTY, originalSR.getHits().getTotalHits(), Float.NaN),
+            null,
+            null,
+            null,
+            false,
+            false,
+            originalSR.getNumReducePhases()
+        );
+        SearchResponse failureSR = new SearchResponse(
+            failureISR,
+            null,
+            originalSR.getTotalShards(),
+            originalSR.getSuccessfulShards(),
+            originalSR.getSkippedShards(),
+            originalSR.getTook().getMillis(),
+            originalSR.getShardFailures(),
+            originalSR.getClusters()
+        );
+        updateException.setStackTrace(new StackTraceElement[0]); // we don't need to store stack traces
```

Comment:
There is no way to set up the limit except copying stack trace array in a new array. I was thinking since this exception is already logged with the full stack trace, it is enough for a stored failure just to have an error's message. WDYT?